File: src/minilake/mcp/tools/warehouses.py
"""MCP tools for SQL warehouses (/api/2.0/sql/warehouses)."""

from typing import Any, Optional

from minilake.mcp.client import MinilakeClient

_PREFIX = "/api/2.0/sql/warehouses"


def register(mcp: Any, client: MinilakeClient) -> None:
    @mcp.tool()
    async def create_warehouse(name: str, cluster_size: Optional[str] = "Small") -> dict[str, Any]:
        """Create a SQL warehouse.

        Each warehouse gets its own DuckDB connection. `cluster_size` is recorded but has
        no effect on performance — there is no real compute to size. You usually do not
        need this: run_sql provisions a shared warehouse on its own.
        """
        return await client.post(_PREFIX, json={"name": name, "cluster_size": cluster_size})

    @mcp.tool()
    async def list_warehouses() -> dict[str, Any]:
        """List all SQL warehouses."""
        return await client.get(_PREFIX)

    @mcp.tool()
    async def get_warehouse(warehouse_id: str) -> dict[str, Any]:
        """Get one warehouse's details by id."""
        return await client.get(f"{_PREFIX}/{warehouse_id}")

    @mcp.tool()
    async def start_warehouse(warehouse_id: str) -> dict[str, Any]:
        """Mark a warehouse RUNNING.

        State flag only — statements execute regardless of warehouse state, so this exists
        for parity with client code that waits for RUNNING.
        """
        return await client.post(f"{_PREFIX}/{warehouse_id}/start")

    @mcp.tool()
    async def stop_warehouse(warehouse_id: str) -> dict[str, Any]:
        """Mark a warehouse STOPPED. State flag only; queries still run afterwards."""
        return await client.post(f"{_PREFIX}/{warehouse_id}/stop")

    @mcp.tool()
    async def delete_warehouse(warehouse_id: str) -> dict[str, Any]:
        """Delete a warehouse and close its DuckDB connection."""
        return await client.delete(f"{_PREFIX}/{warehouse_id}")

File: ui/src/lib/api/warehouses.ts
import { apiFetch } from "./client";
import type { Warehouse } from "./types";

const SQL = "/api/2.0/sql";

export async function listWarehouses(): Promise<Warehouse[]> {
  const data = await apiFetch<{ warehouses?: Warehouse[] }>(`${SQL}/warehouses`);
  return data.warehouses ?? [];
}

export function getWarehouse(id: string): Promise<Warehouse> {
  return apiFetch<Warehouse>(`${SQL}/warehouses/${id}`);
}

/** Create a warehouse. The API returns only `{id}` — re-fetch for the rest. */
export async function createWarehouse(name: string, clusterSize = "Small"): Promise<Warehouse> {
  const { id } = await apiFetch<{ id: string }>(`${SQL}/warehouses`, {
    method: "POST",
    body: { name, cluster_size: clusterSize },
  });
  return getWarehouse(id);
}

export function startWarehouse(id: string): Promise<Warehouse> {
  return apiFetch<Warehouse>(`${SQL}/warehouses/${id}/start`, { method: "POST" });
}

export function stopWarehouse(id: string): Promise<Warehouse> {
  return apiFetch<Warehouse>(`${SQL}/warehouses/${id}/stop`, { method: "POST" });
}

export function deleteWarehouse(id: string): Promise<unknown> {
  return apiFetch(`${SQL}/warehouses/${id}`, { method: "DELETE" });
}
